File: src/test/java/test/java/SignatureTest.java
package test.java;

import me.saro.dat.DatUtils;
import me.saro.dat.signature.DatSignature;
import me.saro.dat.signature.DatSignatureAlgorithm;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class SignatureTest {

    public void unit(DatSignatureAlgorithm alg) {
        String tag = "Signature " + alg;
        byte[] body = DatUtils.generateRandomBase62(100).getBytes();
        DatSignature signatureKey = DatSignature.generate(alg);
        DatSignature signatureKeyFail = DatSignature.generate(alg);
        byte[] keyBytes = signatureKey.exportKey(false);
        byte[] verifyingKeyBytes = signatureKey.exportKey(signatureKey.supportVerifyOnly());
        DatSignature signatureKeyFrom = DatSignature.fromKey(alg, keyBytes);
        DatSignature verifyKeyFrom = DatSignature.fromKey(alg, verifyingKeyBytes);

        var sign = signatureKeyFrom.sign(body);

        assertTrue(signatureKey.verify(body, sign));
        assertTrue(signatureKeyFrom.verify(body, sign));
        assertTrue(verifyKeyFrom.verify(body, sign));
        assertFalse(signatureKeyFail.verify(body, sign));
        System.out.println(tag + " PASS : " + DatUtils.encodeBase64Url(keyBytes));
    }

    @Test
    public void test() {
        for (var algorithm : DatSignatureAlgorithm.getEntries()) {
            for (var i = 0; i < 20; i++) {
                unit(algorithm);
            }
        }
    }
}

File: src/test/kotlin/test/kt/SignatureTest.kt
package test.kt

import me.saro.dat.DatUtils.Companion.encodeBase64Url
import me.saro.dat.DatUtils.Companion.generateRandomBase62
import me.saro.dat.signature.DatSignature.Companion.fromKey
import me.saro.dat.signature.DatSignature.Companion.generate
import me.saro.dat.signature.DatSignatureAlgorithm
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class SignatureTest {
    fun unit(alg: DatSignatureAlgorithm) {
        val tag = "Signature " + alg
        val body = generateRandomBase62(100).toByteArray()
        val signatureKey = generate(alg)
        val signatureKeyFail = generate(alg)
        val keyBytes = signatureKey.exportKey(false)
        val verifyingKeyBytes = signatureKey.exportKey(signatureKey.supportVerifyOnly())
        val signatureKeyFrom = fromKey(alg, keyBytes)
        val verifyKeyFrom = fromKey(alg, verifyingKeyBytes)

        val sign = signatureKeyFrom.sign(body)

        Assertions.assertTrue(signatureKey.verify(body, sign))
        Assertions.assertTrue(signatureKeyFrom.verify(body, sign))
        Assertions.assertTrue(verifyKeyFrom.verify(body, sign))
        Assertions.assertFalse(signatureKeyFail.verify(body, sign))
        println(tag + " PASS : " + encodeBase64Url(keyBytes))
    }

    @Test
    fun test() {
        for (algorithm in DatSignatureAlgorithm.entries) {
            for (i in 0..19) {
                unit(algorithm)
            }
        }
    }
}
